[feat] Add support for the LSF scheduler and LLNL's `lrun` parallel launcher

rngoodner opened the pull request:
This adds basic support for LSF and lrun. LSF is a scheduler and lrun is a launcher used on the Lawrence Livermore National Laboratory Sierra Systems. I have used this implementation to run tests on the Lassen cluster successfully.

I followed the sge scheduler implementation that worked off the pbs implementation as an example.

Closes #1055

pep8speaks commented:
Hello @rngoodner, Thank you for submitting the Pull Request!





* In the file [`reframe/core/schedulers/lsf.py`](https://github.com/eth-cscs/reframe/blob/a07b78c77cb8a7a044d91a9456c22091f5a05e5c/reframe/core/schedulers/lsf.py):

> [Line 34:1](https://github.com/eth-cscs/reframe/blob/a07b78c77cb8a7a044d91a9456c22091f5a05e5c/reframe/core/schedulers/lsf.py#L34): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

* In the file [`reframe/core/schedulers/registry.py`](https://github.com/eth-cscs/reframe/blob/a07b78c77cb8a7a044d91a9456c22091f5a05e5c/reframe/core/schedulers/registry.py):

> [Line 43:1](https://github.com/eth-cscs/reframe/blob/a07b78c77cb8a7a044d91a9456c22091f5a05e5c/reframe/core/schedulers/registry.py#L43): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file



* In the file [`unittests/test_schedulers.py`](https://github.com/eth-cscs/reframe/blob/a07b78c77cb8a7a044d91a9456c22091f5a05e5c/unittests/test_schedulers.py):

> [Line 29:80](https://github.com/eth-cscs/reframe/blob/a07b78c77cb8a7a044d91a9456c22091f5a05e5c/unittests/test_schedulers.py#L29): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 134:1](https://github.com/eth-cscs/reframe/blob/a07b78c77cb8a7a044d91a9456c22091f5a05e5c/unittests/test_schedulers.py#L134): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

Do see the ReFrame [Coding Style Guide](https://github.com/eth-cscs/reframe/wiki/coding-style-guide)

teojgo commented:
Ok to test

vkarak commented on reframe/core/launchers/mpi.py:
```diff
@@ -124,3 +124,26 @@ def command(self, job):
+class LrunLauncher(JobLauncher):
+    '''LLNL's custom parallel job launcher.'''
+    def command(self, job):
+        num_tasks = job.num_tasks or 1
```
I'm not sure that this is needed. `job.num_tasks` is always > 0. If it is set to zero, then the scheduler will try to do a flexible allocation if supported.

rngoodner replied:
done

vkarak commented on reframe/core/launchers/mpi.py:
```diff
@@ -124,3 +124,26 @@ def command(self, job):
+    def command(self, job):
+        num_tasks = job.num_tasks or 1
+        num_tasks_per_node = job.num_tasks_per_node or 1
+        num_nodes = int(num_tasks // num_tasks_per_node)
```
The `//` is by definition the integer division and both operands are guaranteed to be integers:

```suggestion
        num_nodes = num_tasks // num_tasks_per_node
```

rngoodner replied:
done

vkarak commented on reframe/core/launchers/mpi.py:
```diff
@@ -124,3 +124,26 @@ def command(self, job):
+@register_launcher('lrun-gpu')
+class LrungpuLauncher(JobLauncher):
+    '''LLNL's custom parallel job launcher w/ CUDA aware Spectum MPI.'''
+    def command(self, job):
+        num_tasks = job.num_tasks or 1
+        num_tasks_per_node = job.num_tasks_per_node or 1
+        num_nodes = int(num_tasks // num_tasks_per_node)
+        return ['lrun', '-M "-gpu"',
+                '-N', str(num_nodes),
+                '-T', str(num_tasks_per_node)]
```
You could possibly inherit from `LrunLauncher()` and implement `command()` as follows:

```suggestion
    def command(self, job):
        return super().command(job) + ['-M "-gpu"']
```

rngoodner replied:
done

vkarak commented on reframe/core/schedulers/lsf.py:
```diff
@@ -0,0 +1,140 @@
+        )
+
+    def emit_preamble(self, job):
+        num_tasks = job.num_tasks or 1
```
See my comment above:

```suggestion
        num_tasks = job.num_tasks
```

rngoodner replied:
done

vkarak commented on reframe/core/schedulers/lsf.py:
```diff
@@ -0,0 +1,140 @@
+        if not jobs:
+            return
+
+        completed = osext.run_command(
+            f'bjobs -noheader {" ".join(job.jobid for job in jobs)}')
+        if completed.returncode != 0:
+            raise JobSchedulerError(
+                f'bjobs failed with exit code {completed.returncode} '
+                f'(standard error follows):\n{completed.stderr}'
```
This should be enough:

```suggestion
        completed = _run_strict(
            f'bjobs -noheader {" ".join(job.jobid for job in jobs)}'
        )
```

rngoodner replied:
done

vkarak commented on reframe/core/schedulers/lsf.py:
```diff
@@ -0,0 +1,140 @@
+                job_status[job_match['jobid']] = job_match['status']
+
+        for job in jobs:
+            # job id not found
```
Move these comments inside the corresponding `if`s.

vkarak commented on reframe/core/schedulers/lsf.py:
```diff
@@ -0,0 +1,140 @@
+                job._completed = True
+
+            # job done
+            elif job_status[job.jobid] in ['DONE', 'EXIT']:
```
```suggestion
            elif job_status[job.jobid] in ('DONE', 'EXIT'):
```

rngoodner replied:
done

vkarak reviewed (changes requested):
Thanks a lot @rngoodner for your PR! It looks very good overall. I have only some minor comments regarding the implementation. We would also need the corresponding entries about the added scheduler and parallel launchers in the docs.

vkarak commented:
@AnibalMG could you test if this LSF backend works for your case?

rngoodner commented:
@vkarak I have addressed all of your comments and also updated the docs. I have also re-tested on the LLNL Lassen cluster and everything is still working for my needs.

vkarak reviewed (approved):
Lgtm! I just did a couple of minor style fixes.
